src/main/java/com/solace/samples/jcsmp/snippets/HowToConfigureSessionUsingOpenIdConnect.java

```diff
@@ -16,61 +16,117 @@
+import com.solacesystems.jcsmp.*;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import com.solacesystems.jcsmp.InvalidPropertiesException;
-import com.solacesystems.jcsmp.JCSMPException;
-import com.solacesystems.jcsmp.JCSMPFactory;
-import com.solacesystems.jcsmp.JCSMPSession;
-import com.solacesystems.jcsmp.SessionEvent;
-import com.solacesystems.jcsmp.SessionEventArgs;
-import com.solacesystems.jcsmp.SessionEventHandler;
+
+import java.time.Instant;
 
 /**
  * Demonstrates how to configure a Session using properties supporting OpenId
  * Connect authentication
  */
 public class HowToConfigureSessionUsingOpenIdConnect {
-  private JCSMPSession session = null;
-
-  // remember to add log4j2.xml to your classpath
-  private static final Logger logger = LogManager.getLogger(); // log4j2, but could also use SLF4J, JCL, etc.
-
-  public void createOIDCSession(JCSMPProperties sessionProperties, String idToken, String accessTokenOptional)
-      throws InvalidPropertiesException {
-    sessionProperties.setProperty(JCSMPProperties.AUTHENTICATION_SCHEME, JCSMPProperties.AUTHENTICATION_SCHEME_OAUTH2);
-    sessionProperties.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, accessTokenOptional);
-    sessionProperties.setProperty(JCSMPProperties.OIDC_ID_TOKEN, idToken);
-
-    // When the OAuth token has expired, the client will be disconnected, and
-    // will try to reconnect. Refresh the token before it reconnects.
-    session = JCSMPFactory.onlyInstance().createSession(sessionProperties, null,
-        new SessionEventHandler() {
-          @Override
-          public void handleEvent(SessionEventArgs event) {
-            if (event.getEvent() == SessionEvent.RECONNECTING) {
-              logger.info("Session Reconnecting - Refresh OAuth tokens");
-
-              try {
-                session.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, refreshAccessToken());
-                session.setProperty(JCSMPProperties.OIDC_ID_TOKEN, refreshOAuthIDToken());
-              } catch (JCSMPException e) {
-                logger.error("Failed to refresh OAuth tokens");
-              }
+    private JCSMPSession session = null;
+
+    // remember to add log4j2.xml to your classpath
+    private static final Logger logger = LogManager.getLogger(); // log4j2, but could also use SLF4J, JCL, etc.
+
+    public void createOIDCSession(JCSMPProperties sessionProperties, String idToken, String accessTokenOptional, String issuerId) throws InvalidPropertiesException {
+        sessionProperties.setProperty(JCSMPProperties.AUTHENTICATION_SCHEME, JCSMPProperties.AUTHENTICATION_SCHEME_OAUTH2);
+        sessionProperties.setProperty(JCSMPProperties.OAUTH2_ACCESS_TOKEN, accessTokenOptional);
+        sessionProperties.setProperty(JCSMPProperties.OIDC_ID_TOKEN, idToken);
+
+        // When the OAuth token has expired, the client will be disconnected, and
+        // will try to reconnect. Refresh the token before it reconnects.

```

Comment:
shouldn't we be checking to see if the token is close to being expired and refreshing it before expiry?

Reply:
No, the new token will be pickedup only when the client tries to reconnect after an expired token exception.
So, the best practice as per the field (SEs and PSGs) is to catch the exception and handle the token refresh accordingly.